Instrumenter triggers on attempt to place EObjectHint on record.

...it considers records as non-base classes, for which placing the hint is prohibited:

```
[ERROR] XXX: It's prohibited to place EObjectHint on inherited classes
```

--- a/oo-equivalence/equivalence-codegen/src/main/java/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClass.java
+++ b/oo-equivalence/equivalence-codegen/src/main/java/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClass.java
@@ -1,0 +1,45 @@
+/*-
+ * ===========================================================================
+ * equivalence-codegen
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Copyright (C) 2019 - 2024 Kapralov Sergey
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * ============================================================================
+ */
+package com.pragmaticobjects.oo.equivalence.codegen.matchers;
+
+import net.bytebuddy.matcher.ElementMatchers;
+
+/**
+ * Match class that is inheritor. In other words, match a class, parent of which is something different from
+ * the {@link Object}. Typically, for these classes, EObjectHint is disallowed, with some exceptions out of the rule.
+ *
+ * The class exists to cover these exceptions. For example, records are formally allowed to be hinted. Yet, they are not directly
+ * subclassed from {@link Object}.
+ */
+public class MatchInheritedClass extends MatchSuperClass {
+    public MatchInheritedClass() {
+        super(
+            ElementMatchers.not(
+                ElementMatchers.is(Object.class)
+            )
+        );
+    }
+}

--- a/oo-equivalence/equivalence-codegen/src/main/java/com/pragmaticobjects/oo/equivalence/codegen/stage/StandardInstrumentationStage.java
+++ b/oo-equivalence/equivalence-codegen/src/main/java/com/pragmaticobjects/oo/equivalence/codegen/stage/StandardInstrumentationStage.java
@@ -51,11 +51,7 @@ public class StandardInstrumentationStage extends SequenceStage {
             // Pre-checks
             new ByteBuddyValidationStage(
                 "It's prohibited to place EObjectHint on inherited classes",
-                new MatchSuperClass(
-                    ElementMatchers.not(
-                        ElementMatchers.is(Object.class)
-                    )
-                ),
+                new MatchInheritedClass(),
                 ElementMatchers.not(
                     ElementMatchers.isAnnotatedWith(EObjectHint.class)
                 )

--- a/oo-equivalence/equivalence-codegen/src/main/java16/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClass.java
+++ b/oo-equivalence/equivalence-codegen/src/main/java16/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClass.java
@@ -1,0 +1,50 @@
+/*-
+ * ===========================================================================
+ * equivalence-codegen
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Copyright (C) 2019 - 2024 Kapralov Sergey
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * ============================================================================
+ */
+package com.pragmaticobjects.oo.equivalence.codegen.matchers;
+
+import net.bytebuddy.matcher.ElementMatchers;
+
+/**
+ * Match class that is inheritor. In other words, match a class, parent of which is something different from
+ * the {@link Object}. Typically, for these classes, EObjectHint is disallowed, with some exceptions out of the rule.
+ *
+ * The class exists to cover these exceptions. For example, records are formally allowed to be hinted. Yet, they are not directly
+ * subclassed from {@link Object}.
+ */
+public class MatchInheritedClass extends MatchSuperClass {
+    public MatchInheritedClass() {
+        super(
+            new ConjunctionMatcher<>(
+                ElementMatchers.not(
+                    ElementMatchers.is(Object.class)
+                ),
+                ElementMatchers.not(
+                    ElementMatchers.is(Record.class)
+                )
+            )
+        );
+    }
+}

--- a/oo-equivalence/equivalence-codegen/src/test/java/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClassTest.java
+++ b/oo-equivalence/equivalence-codegen/src/test/java/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClassTest.java
@@ -1,0 +1,55 @@
+/*-
+ * ===========================================================================
+ * equivalence-codegen
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Copyright (C) 2019 - 2024 Kapralov Sergey
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * ============================================================================
+ */
+package com.pragmaticobjects.oo.equivalence.codegen.matchers;
+
+import com.pragmaticobjects.oo.equivalence.assertions.TestCase;
+import com.pragmaticobjects.oo.equivalence.assertions.TestsSuite;
+
+public class MatchInheritedClassTest extends TestsSuite {
+    public MatchInheritedClassTest() {
+        super(
+            new TestCase(
+                "base class case",
+                new AssertThatTypeDoesNotMatch(
+                    BaseClass.class,
+                    new MatchInheritedClass()
+                )
+            ),
+            new TestCase(
+                "inheritor class case",
+                new AssertThatTypeMatches(
+                    InheritedClass.class,
+                    new MatchInheritedClass()
+                )
+            )
+        );
+    }
+}
+
+class BaseClass {}
+
+class InheritedClass extends BaseClass {}
+

--- a/oo-equivalence/equivalence-codegen/src/test/java16/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClassTest.java
+++ b/oo-equivalence/equivalence-codegen/src/test/java16/com/pragmaticobjects/oo/equivalence/codegen/matchers/MatchInheritedClassTest.java
@@ -1,0 +1,61 @@
+/*-
+ * ===========================================================================
+ * equivalence-codegen
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Copyright (C) 2019 - 2024 Kapralov Sergey
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * ============================================================================
+ */
+package com.pragmaticobjects.oo.equivalence.codegen.matchers;
+
+import com.pragmaticobjects.oo.equivalence.assertions.TestCase;
+import com.pragmaticobjects.oo.equivalence.assertions.TestsSuite;
+
+public class MatchInheritedClassTest extends TestsSuite {
+    public MatchInheritedClassTest() {
+        super(
+            new TestCase(
+                "base class case",
+                new AssertThatTypeDoesNotMatch(
+                    BaseClass.class,
+                    new MatchInheritedClass()
+                )
+            ),
+            new TestCase(
+                "inheritor class case",
+                new AssertThatTypeMatches(
+                    InheritedClass.class,
+                    new MatchInheritedClass()
+                )
+            )/*,
+            TODO: uncomment this test when tests classloding with multi-release structure is fixed
+            new TestCase(
+                "records are not considered inherited",
+                new AssertThatTypeDoesNotMatch(
+                    SomeRecord.class,
+                    new MatchInheritedClass()
+                )
+            )
+            */
+        );
+    }
+}
+
+record SomeRecord() {}

--- a/oo-equivalence/equivalence-itests/test-cases/src/test/java16/com/pragmaticobjects/oo/equivalence/itests/records/TestHintedRecords.java
+++ b/oo-equivalence/equivalence-itests/test-cases/src/test/java16/com/pragmaticobjects/oo/equivalence/itests/records/TestHintedRecords.java
@@ -1,0 +1,48 @@
+/*-
+ * ===========================================================================
+ * equivalence-itests.test-cases
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Copyright (C) 2019 - 2024 Kapralov Sergey
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * ============================================================================
+ */
+package com.pragmaticobjects.oo.equivalence.itests.records;
+
+import com.pragmaticobjects.oo.equivalence.assertions.AssertNotInstanceOf;
+import com.pragmaticobjects.oo.equivalence.assertions.AssertSubtypeOf;
+import com.pragmaticobjects.oo.equivalence.assertions.TestCase;
+import com.pragmaticobjects.oo.equivalence.assertions.TestsSuite;
+import com.pragmaticobjects.oo.equivalence.base.EquivalenceCompliant;
+import com.pragmaticobjects.oo.equivalence.itests.classes.IgnoredClass;
+
+/**
+ *
+ * @author skapral
+ */
+public class TestHintedRecords extends TestsSuite {
+    public TestHintedRecords() {
+        super(
+            new TestCase(
+                "explicit ignoring of a record",
+                new AssertNotInstanceOf(IgnoredRecord.class, EquivalenceCompliant.class)
+            )
+        );
+    }
+}

--- a/oo-equivalence/equivalence-itests/test-module2/src/main/java16/com/pragmaticobjects/oo/equivalence/itests/records/IgnoredRecord.java
+++ b/oo-equivalence/equivalence-itests/test-module2/src/main/java16/com/pragmaticobjects/oo/equivalence/itests/records/IgnoredRecord.java
@@ -1,0 +1,32 @@
+/*-
+ * ===========================================================================
+ * equivalence-itests.test-module2
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Copyright (C) 2019 - 2024 Kapralov Sergey
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * ============================================================================
+ */
+package com.pragmaticobjects.oo.equivalence.itests.records;
+
+import com.pragmaticobjects.oo.equivalence.base.EObjectHint;
+
+@EObjectHint(enabled = false)
+public record IgnoredRecord(int id) {
+}

